sd-journal: properly handle inotify queue overflow (#1540538)

msekletar opened the pull request.

portante committed:
journalctl: Periodically call sd_journal_process in journalctl

If `journalctl` take a long time to process messages, and during that
time journal file rotation occurs, a `journalctl` client will keep
those rotated files open until it calls `sd_journal_process()`, which
typically happens as a result of calling `sd_journal_wait()` below in
the "following" case.  By periodically calling `sd_journal_process()`
during the processing loop we shrink the window of time a client
instance has open file descriptors for rotated (deleted) journal
files.

(Lennart: slightly reworked version, that dropped some of the commenting
which was solved otherwise)

(cherry picked from commit ec316d199a13d8db3f6550d60e369893de2fb417)

Related: #1540538

poettering committed:
sd-journal: make sure it's safe to call sd_journal_process() before the first sd_journal_wait()

In that case we have no inotify fd yet, and there's nothing to process
hence. Let's make the call a NOP.

(Previously, without this change we'd end up trying to read off inotify
fd -1, which is quite a problem... 😢)

(cherry picked from commit 10c4d6405f74258ea4fac5db4888c1bf49ad5399)

Related: #1540538

poettering committed:
sd-journal: properly handle inotify queue overflow

This adds proper handling of IN_Q_OVERFLOW: when the inotify queue runs
over we'll reiterate all directories we are looking at. At the same time
we'll mark all files and directories we encounter that way with a
generation counter we first increased. All files and directories not
marked like this are then unloaded.

With this logic we do the best when the inotify queue overflows: we
synchronize our in-memory state again with what's on disk.  This
contains some refactoring of the directory logic, to share more code
between uuid directories and "root" directories and generally make
things a bit more readable by splitting things up into smaller bits.

See: #7998 #8032

(cherry-picked from commit 858749f7312bd0adb5433075a92e1c35a2fb56ac)

Resolves: #1540538

lnykryn commented on src/journal/sd-journal.c:
```diff
@@ -1229,8 +1229,16 @@ static int add_any_file(sd_journal *j, const char *path) {
 
-        if (ordered_hashmap_get(j->files, path))
-                return 0;
+        if (path) {
```
This part looks a bit suspicious:
assert(path);
followed with
if (path) {

lnykryn replied:
Ok, upstream code can take fd, instead of path, so ... meh.

msekletar committed:
sd-journal: when picking up a new file, compare inode/device info with previous open file by same name

Let's make sure we aren't confused if a journal file is replaced by a
different one (for example due to rotation) if we are in a q overflow:
let's compare the inode/device information, and if it changed replace
any open file object as needed.

Fixes: #8198

(cherry-picked from commit 32cb1983ad6f7084ff86e259ff079742a8139719)

[msekleta: this is very slimmed down version of the above commit because
a lot of code from is not applicable to RHEL-7 version]

Related: #1540538

lnykryn reviewed (approved):
LGTM
